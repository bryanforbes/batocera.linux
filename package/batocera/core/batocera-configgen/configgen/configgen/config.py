from __future__ import annotations

import logging
import xml.etree.ElementTree as ET
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, ClassVar, Literal, Self, TypeVar, overload

import yaml

from .batoceraPaths import BATOCERA_SHADERS, DEFAULTS_DIR, ES_SETTINGS, USER_SHADERS
from .utils.missing import MISSING, MissingType

if TYPE_CHECKING:
    from pathlib import Path

    from .cli import Options
    from .settings.unixSettings import UnixSettings

_T = TypeVar('_T')
_TV = TypeVar('_TV')
_FV = TypeVar('_FV')

_logger = logging.getLogger(__name__)


# adapted from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
def _dict_merge(destination: dict[str, Any], source: Mapping[str, Any]) -> None:
    """Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param destination: dict onto which the merge is executed
    :param source: dict merged into destination
    :return: None
    """
    for key, value in source.items():
        if key in destination and isinstance(destination[key], dict) and isinstance(value, Mapping):
            _dict_merge(destination[key], value)
        else:
            destination[key] = value


def _load_defaults(system: str, default_file: Path, default_arch_file: Path, /) -> dict[str, Any]:
    with default_file.open() as f:
        defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader)

    with default_arch_file.open() as f:
        arch_defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader) or {}

    config: dict[str, Any] = defaults.get('default', {})

    if 'default' in arch_defaults:
        _dict_merge(config, arch_defaults['default'])

    if system in defaults:
        _dict_merge(config, defaults[system])

    if system in arch_defaults:
        _dict_merge(config, arch_defaults[system])

    return config


@dataclass(slots=True)
class Config:
    TRUE_VALUES: ClassVar[set[Literal['1', 'true', 'on', 'enabled', True]]] = {'1', 'true', 'on', 'enabled', True}
    MISSING: ClassVar = MISSING

    data: dict[str, Any]

    def __contains__(self, x: object, /) -> bool:
        return x in self.data

    @overload
    def get(self, key: str, default: MissingType = ..., /) -> Any | MissingType: ...

    @overload
    def get(self, key: str, default: _T, /) -> Any | _T: ...

    def get(self, key: str, default: _T | MissingType = MISSING, /) -> Any | _T | MissingType:
        return self.data.get(key, default)

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: None = None) -> bool: ...

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV]) -> _TV | _FV: ...

    def get_bool(
        self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV] | None = None
    ) -> bool | _TV | _FV:
        value = self.get(key)

        if value is MISSING:
            return default

        if isinstance(value, str):
            value = value.lower()

        if return_values is None:
            return value in self.TRUE_VALUES

        return return_values[value not in self.TRUE_VALUES]

    @overload
    def get_str(self, key: str, default: MissingType = ..., /) -> str | MissingType: ...

    @overload
    def get_str(self, key: str, default: str, /) -> str: ...

    def get_str(self, key: str, default: str | MissingType = MISSING, /) -> str | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return str(value)

    @overload
    def get_int(self, key: str, default: MissingType = ..., /) -> int | MissingType: ...

    @overload
    def get_int(self, key: str, default: int, /) -> int: ...

    def get_int(self, key: str, default: int | MissingType = MISSING, /) -> int | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return int(value)

    @overload
    def get_float(self, key: str, default: MissingType = ..., /) -> float | MissingType: ...

    @overload
    def get_float(self, key: str, default: float, /) -> float: ...

    def get_float(self, key: str, default: float | MissingType = MISSING, /) -> float | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return float(value)

    def items(self, /, *, starts_with: str | None = None) -> Iterator[tuple[str, Any]]:
        if starts_with is None:
            yield from self.data.items()
        else:
            starts_with_len = len(starts_with)
            for key, value in self.data.items():
                if key.startswith(starts_with):
                    yield key[starts_with_len:], value


def _get_rom_settings_name(rom: Path, /) -> str:
    # sanitize rule by EmulationStation
    # see FileData::getConfigurationName() on batocera-emulationstation
    rom_settings_name = rom.name.replace('=', '').replace('#', '')

    _logger.info('game settings name: %s', rom_settings_name)

    return rom_settings_name


@dataclass
class SystemConfig(Config):
    @property
    def core(self) -> str:
        return self.data['core']

    @property
    def core_forced(self) -> str:
        return self.data['core-forced']

    @property
    def emulator(self) -> str:
        return self.data['emulator']

    @property
    def emulator_forced(self) -> str:
        return self.data['emulator-forced']

    @property
    def ui_mode(self) -> Literal['Full', 'Kiosk', 'Kid']:
        return self.data['uimode']

    @property
    def show_fps(self) -> bool:
        return self.data['showFPS']

    @property
    def video_mode(self) -> str:
        return self.data.get('videomode', 'default')

    @property
    def use_guns(self) -> bool:
        return self.get_bool('use_guns')

    @property
    def use_wheels(self) -> bool:
        return self.get_bool('use_wheels')

    @classmethod
    def load(cls, options: Options, settings: UnixSettings, /) -> Self:
        defaults = _load_defaults(
            options['system'], DEFAULTS_DIR / 'configgen-defaults.yml', DEFAULTS_DIR / 'configgen-defaults-arch.yml'
        )
        data: dict[str, Any] = {
            'emulator': defaults['emulator'],
            'core': defaults['core'],
        }

        # In the yaml files, the "options" structure is not flat, so we have to flatten it here
        # because the options are flat in batocera.conf to make it easier for end users to edit
        if 'options' in defaults:
            _dict_merge(data, defaults['options'])

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        data.update(settings.get_all_iter('display', keep_name=True, keep_defaults=True))
        data.update(settings.get_all_iter('controllers', keep_name=True))

        global_settings = settings.get_all('global')
        system_settings = settings.get_all(options['system'])
        folder_settings = settings.get_all(f'{options["system"]}.folder["{rom.parent}"]')
        game_settings = settings.get_all(f'{options["system"]}["{rom_settings_name}"]')

        data.update(global_settings)
        data.update(system_settings)
        data.update(folder_settings)
        data.update(game_settings)

        try:
            es_config = ET.parse(ES_SETTINGS)

            # showFPS
            drawframerate_node = es_config.find("./bool[@name='DrawFramerate']")
            drawframerate_value = drawframerate_node.attrib['value'] if drawframerate_node is not None else 'false'
            if drawframerate_value not in ['false', 'true']:
                drawframerate_value = 'false'

            data['showFPS'] = drawframerate_value == 'true'

            # uimode
            uimode_node = es_config.find("./string[@name='UIMode']")
            uimode_value = uimode_node.attrib['value'] if uimode_node is not None else 'Full'
            if uimode_value not in ['Full', 'Kiosk', 'Kid']:
                uimode_value = 'Full'

            data['uimode'] = uimode_value
        except Exception:
            data['showFPS'] = False
            data['uimode'] = 'Full'

        data['emulator-forced'] = (
            'emulator' in global_settings
            or 'emulator' in system_settings
            or 'emulator' in game_settings
            or options['emulator'] is not None
        )
        data['core-forced'] = (
            'core' in global_settings
            or 'core' in system_settings
            or 'core' in game_settings
            or options['core'] is not None
        )

        if options['emulator'] is not None:
            data['emulator'] = options['emulator']

        if options['core'] is not None:
            data['core'] = options['core']

        if 'use_guns' not in data and options['lightgun']:
            data['use_guns'] = True

        if 'use_wheels' not in data and options['wheel']:
            data['use_wheels'] = True

        # network options
        if options['netplaymode'] is not None:
            data['netplay.mode'] = options['netplaymode']

        if options['netplaypass'] is not None:
            data['netplay.password'] = options['netplaypass']

        if options['netplayip'] is not None:
            data['netplay.server.ip'] = options['netplayip']

        if options['netplayport'] is not None:
            data['netplay.server.port'] = options['netplayport']

        if options['netplaysession'] is not None:
            data['netplay.server.session'] = options['netplaysession']

        # autosave arguments
        if options['state_slot'] is not None:
            data['state_slot'] = options['state_slot']

        if options['autosave'] is not None:
            data['autosave'] = options['autosave']

        if options['state_filename'] is not None:
            data['state_filename'] = options['state_filename']

        return cls(data)


@dataclass
class RenderConfig(Config):
    @classmethod
    def load(cls, options: Options, settings: UnixSettings, shader_set: str | MissingType, /) -> Self:
        data: dict[str, Any] = {}

        if shader_set is not cls.MISSING:
            if shader_set == 'none':
                rendering_defaults = BATOCERA_SHADERS / 'configs' / 'rendering-defaults.yml'
            else:
                rendering_defaults = USER_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'
                if not rendering_defaults.exists():
                    rendering_defaults = BATOCERA_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'

            data = _load_defaults(
                options['system'], rendering_defaults, rendering_defaults.with_name('rendering-defaults-arch.yml')
            )

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        # es only allow to update system and game settings in fact for the moment

        # for compatibility with earlier Batocera versions, let's keep -renderer
        # but it should be reviewed when we refactor configgen (to Python3?)
        # so that we can fetch them from system.shader without -renderer
        data.update(settings.get_all_iter(f'{options["system"]}-renderer'))
        data.update(settings.get_all_iter(f'{options["system"]}["{rom_settings_name}"]-renderer'))

        return cls(data)
