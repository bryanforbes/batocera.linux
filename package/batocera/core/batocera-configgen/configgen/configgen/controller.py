from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
