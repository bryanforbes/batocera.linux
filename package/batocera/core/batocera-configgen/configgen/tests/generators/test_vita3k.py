from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, Any

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS, SAVES
from configgen.config import SystemConfig
from configgen.generators.vita3k.vita3kGenerator import Vita3kGenerator
from tests.generators.base import GeneratorBaseTest

if TYPE_CHECKING:
    from pyfakefs.fake_filesystem import FakeFilesystem
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controllers
    from configgen.Emulator import Emulator


@pytest.mark.usefixtures('fs')
class TestVita3kGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[Vita3kGenerator]:
        return Vita3kGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'psvita'

    @pytest.fixture
    def emulator(self) -> str:
        return 'vita3k'

    @pytest.mark.parametrize(
        ('mock_system_config', 'result'),
        [({}, True), ({'vita3k_show_pointer': '1'}, True), ({'vita3k_show_pointer': '0'}, False)],
        ids=str,
    )
    def test_get_mouse_mode(  # pyright: ignore
        self, generator: Vita3kGenerator, mock_system_config: dict[str, Any], result: bool
    ) -> None:
        assert generator.getMouseMode(SystemConfig(mock_system_config), Path()) == result

    def test_get_in_game_ratio(self, generator: Vita3kGenerator) -> None:  # pyright: ignore
        assert generator.getInGameRatio(SystemConfig({}), {'width': 0, 'height': 0}, Path()) == 16 / 9

    def test_generate(
        self,
        generator: Vita3kGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'psvita' / 'rom.zip',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (CONFIGS / 'vita3k' / 'config.yml').read_text() == snapshot(name='config')

    def test_generate_existing(
        self,
        generator: Vita3kGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file(
            CONFIGS / 'vita3k' / 'config.yml',
            contents="""
pref-path: /some/path
backend-renderer: Vulkan
resolution-multiplier: 3
enable-fxaa: 'true'
v-sync: 'false'
anisotropic-filtering: 8
enable-linear-filter: 'true'
foo: bar
""",
        )
        generator.generate(
            mock_system,
            ROMS / 'psvita' / 'rom.zip',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'vita3k' / 'config.yml').read_text() == snapshot

    def test_generate_existing_empty(
        self,
        generator: Vita3kGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file(CONFIGS / 'vita3k' / 'config.yml')

        generator.generate(
            mock_system,
            ROMS / 'psvita' / 'rom.zip',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'vita3k' / 'config.yml').read_text() == snapshot

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'vita3k_gfxbackend': 'Vulkan'},
            {'vita3k_resolution': '3'},
            {'vita3k_fxaa': '0'},
            {'vita3k_fxaa': '1'},
            {'vita3k_vsync': '0'},
            {'vita3k_vsync': '1'},
            {'vita3k_anisotropic': '8'},
            {'vita3k_linear': '0'},
            {'vita3k_linear': '1'},
            {'vita3k_surface': '0'},
            {'vita3k_surface': '1'},
        ],
        ids=str,
    )
    def test_generate_config(
        self,
        generator: Vita3kGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'psvita' / 'rom.zip',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'vita3k' / 'config.yml').read_text() == snapshot

    def test_generate_move_saves(
        self,
        generator: Vita3kGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
    ) -> None:
        fs.create_dir(CONFIGS / 'vita3k' / 'ux0')
        fs.create_dir(CONFIGS / 'vita3k' / 'data')
        fs.create_dir(CONFIGS / 'vita3k' / 'lang')
        fs.create_dir(CONFIGS / 'vita3k' / 'shaders-builtin')
        fs.create_dir(CONFIGS / 'vita3k' / 'test')
        fs.create_file(CONFIGS / 'vita3k' / 'foo.txt')

        generator.generate(
            mock_system,
            ROMS / 'psvita' / 'rom.zip',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert not (CONFIGS / 'vita3k' / 'ux0').exists()
        assert (CONFIGS / 'vita3k' / 'foo.txt').exists()
        assert (CONFIGS / 'vita3k' / 'data').exists()
        assert (CONFIGS / 'vita3k' / 'lang').exists()
        assert (CONFIGS / 'vita3k' / 'shaders-builtin').exists()
        assert not (CONFIGS / 'vita3k' / 'test').exists()
        assert (SAVES / 'psvita' / 'ux0').is_dir()
        assert not (SAVES / 'psvita' / 'foo.txt').exists()
        assert not (SAVES / 'psvita' / 'data').exists()
        assert not (SAVES / 'psvita' / 'lang').exists()
        assert not (SAVES / 'psvita' / 'shaders-builtin').exists()
        assert (SAVES / 'psvita' / 'test').is_dir()

    def test_generate_read_only(
        self,
        generator: Vita3kGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_dir(SAVES / 'psvita' / 'ux0' / 'app' / 'foo')

        assert (
            generator.generate(
                mock_system,
                ROMS / 'psvita' / 'rom name[foo].zip',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
