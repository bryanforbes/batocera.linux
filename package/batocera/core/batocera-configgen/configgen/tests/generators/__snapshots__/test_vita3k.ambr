# serializer version: 1
# name: TestVita3kGenerator.test_generate
  tuple(
    list([
      '/usr/bin/vita3k/Vita3K',
      '-F',
      '-w',
      '-f',
      '-c',
      '/userdata/system/configs/vita3k/config.yml',
      '/userdata/roms/psvita/rom.zip',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
      'XDG_CACHE_HOME': '/userdata/system/cache',
      'XDG_CONFIG_HOME': '/userdata/system/configs',
      'XDG_DATA_HOME': '/userdata/saves',
    }),
  )
# ---
# name: TestVita3kGenerator.test_generate[config]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_anisotropic': '8'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 8
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_fxaa': '0'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_fxaa': '1'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'true'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_gfxbackend': 'Vulkan'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: Vulkan
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_linear': '0'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_linear': '1'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'true'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_resolution': '3'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 3
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_surface': '0'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'false'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_surface': '1'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_vsync': '0'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'false'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_config[{'vita3k_vsync': '1'}]
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_existing
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  foo: bar
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_existing_empty
  '''
  ---
  pref-path: /userdata/saves/psvita
  backend-renderer: OpenGL
  resolution-multiplier: 1
  enable-fxaa: 'false'
  v-sync: 'true'
  anisotropic-filtering: 1
  enable-linear-filter: 'false'
  disable-surface-sync: 'true'
  ...
  
  '''
# ---
# name: TestVita3kGenerator.test_generate_read_only
  tuple(
    list([
      '/usr/bin/vita3k/Vita3K',
      '-F',
      '-w',
      '-f',
      '-c',
      '/userdata/system/configs/vita3k/config.yml',
      '-r',
      'foo',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
      'XDG_CACHE_HOME': '/userdata/system/cache',
      'XDG_CONFIG_HOME': '/userdata/system/configs',
      'XDG_DATA_HOME': '/userdata/saves',
    }),
  )
# ---
# name: TestVita3kGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTCTRL',
        'KEY_Q',
      ]),
      'menu': 'KEY_ENTER',
      'pause': 'KEY_ENTER',
    }),
    'name': 'vita3k',
  })
# ---
